### PR TITLE
adcs-66 implement variable timestep

### PR DESCRIPTION
**Pull request by Aidan-Sheedy**
CHANGES
- Added variable timestep algorithm in Simulator.cpp
- Added variable timestep parameters to the YAMLs

REASON
Improves performance testing while retaining accuracy of low timesteps.

TESTING
Tests conducted using the second controller yaml file.

GITHUB LINK
https://github.com/queens-satellite-team/adcs/issues/66

Signed-off-by: Aidan Sheedy <Aidan.P.Sheedy@gmail.com>

### CODE REVIEW

File: csdc-6/adcs-simulation/cpp/src/Simulator.cpp [hunk 1]
```diff
@@ -30,6 +30,10 @@ Simulator::Simulator(Messenger *messenger)
     timestamp t(config.GetTimestepInMilliSeconds(),0);
     this->timestep_length = t;
     this->last_called = -1;
+
+    this->variableTimestep = config.GetTimestepDecision();
```
Comment (Aidan-Sheedy) [inline csdc-6/adcs-simulation/cpp/src/Simulator.cpp:34]: Everything from line 28 -> line 36 should not be in Simulator.cpp.

These should be moved to the UI.cpp, in the function "get_sim_config". The parameters used should be added to the structure "sim_config".

File: csdc-6/adcs-simulation/cpp/src/Simulator.cpp [hunk 2]
```diff
@@ -49,6 +53,25 @@ timestamp Simulator::update_simulation() {
     return this->simulation_time;
 }
 
+timestamp Simulator::determine_timestep() 
+{
+    //2*max error is defined as 2*0.005 degrees = 0.01 degrees
+    if (this->variableTimestep == true){
+        float calculated_timestep = (0.01*M_PI/180)/(this->system_vals.satellite.alpha_b.cwiseAbs().maxCoeff())*1000;
```
Comment (Aidan-Sheedy) [inline csdc-6/adcs-simulation/cpp/src/Simulator.cpp:60]: What are these values? What do they mean? please add them as a static const member of Simulator.cpp with an appropraite name
ie

static const float name = 0.01 * M_PI / 180

File: csdc-6/adcs-simulation/cpp/src/Simulator.cpp [hunk 3]
```diff
@@ -49,6 +53,25 @@ timestamp Simulator::update_simulation() {
...
 
+timestamp Simulator::determine_timestep() 
+{
+    //2*max error is defined as 2*0.005 degrees = 0.01 degrees
+    if (this->variableTimestep == true){
+        float calculated_timestep = (0.01*M_PI/180)/(this->system_vals.satellite.alpha_b.cwiseAbs().maxCoeff())*1000;
+        timestamp t(calculated_timestep,0);
+        this->timestep_length = t;
+        if (((float) timestep_length > maxStep/1000)) {
+            timestamp t(20,0);
```
Comment (Aidan-Sheedy) [inline csdc-6/adcs-simulation/cpp/src/Simulator.cpp:64]: Same comment, where is the 20 coming from? if it's a default value, it should be a const member of the Simulator.cpp

File: csdc-6/adcs-simulation/cpp/src/Simulator.cpp [hunk 4]
```diff
@@ -49,6 +53,25 @@ timestamp Simulator::update_simulation() {
...
+            this->timestep_length = t;
+        }else if ((float) timestep_length < minStep/1000) {
+            timestamp t(1,0);
+            this->timestep_length = t;
+        }
+    }
+
+    return timestep_length;
```
Comment (Aidan-Sheedy) [inline csdc-6/adcs-simulation/cpp/src/Simulator.cpp:72]: This doesn't need a return value, you're already changing "this->timestep_length" in the function.

**Review by justinpaoli — APPROVED** [review]

**Review by ldeloe — APPROVED** [review]

**Review by justinpaoli — APPROVED** [review]

**Review by JonathanFawcett — CHANGES_REQUESTED** [review]
Hacky implementation of webbrowser to open plots, please replace with API between WSL and Windows

**Pull request merged**